File: web/src/triage.test.ts
import { describe, expect, it } from "vitest";
import type { Card } from "./providers/types";
import {
  anchorFor,
  byPile,
  deferred,
  needsTriage,
  orderWith,
  pileRank,
  placedAhead,
  placedIn,
  reachOf,
  weeksCovered,
  broughtBack,
  ordersWithinCell,
} from "./triage";

const card = (over: Partial<Card> = {}): Card =>
  ({ itemId: "c1", title: "A card", assignees: [], ...over }) as Card;

describe("needsTriage", () => {
  it("wants a week for an open card of its own", () => {
    expect(needsTriage(card())).toBe(true);
  });

  it("asks nothing of a card that already has one", () => {
    expect(needsTriage(card({ week: "2026-08-31" }))).toBe(false);
  });

  it("asks nothing of a review — it follows the card it reviews", () => {
    // A review is not work anybody schedules: it lives and dies with its
    // original, so putting it in the strip only asked for a decision that
    // is not the reader's to make.
    expect(needsTriage(card({ reviewOf: "c9" }))).toBe(false);
  });

  it("asks nothing of a subtask — it follows its parent", () => {
    expect(needsTriage(card({ parent: "c9" }))).toBe(false);
  });

  it("asks nothing of a card on a personal board", () => {
    expect(needsTriage(card({ domain: "~kvaps" }))).toBe(false);
  });

  it("asks nothing of a card sent to review — it waits on a reviewer", () => {
    // Its work is done; a week is not what it is missing, and asking for one
    // asks the reader to decide something nobody is waiting on them for.
    expect(needsTriage(card({ stage: "review", progress: 85 }))).toBe(false);
  });

  it("still asks about a locked card — that is neither done nor in review", () => {
    expect(needsTriage(card({ stage: "locked" }))).toBe(true);
  });

  it("asks nothing of work already finished", () => {
    expect(needsTriage(card({ stage: "done" }))).toBe(false);
    expect(needsTriage(card({ progress: 100 }))).toBe(false);
  });

  it("still asks about a card that is merely under way", () => {
    expect(needsTriage(card({ progress: 40 }))).toBe(true);
  });

  it("does not take a day on the board for an answer", () => {
    // The day's planning put it there, not the week's: nobody has said how
    // long the work takes, which is the whole of the question.
    expect(needsTriage(card({ day: "2026-09-02" }))).toBe(true);
  });

  it("asks nothing of a parked card — the answer was 'not now'", () => {
    // The strip is the question "when?", and parking a card ANSWERS it: not
    // now. Leaving it in the strip would ask again every morning, which is
    // how an inbox stops meaning anything.
    expect(needsTriage(card({ parked: true }))).toBe(false);
  });

  it("asks again the moment a card comes off its shelf", () => {
    expect(needsTriage(card({ parked: false }))).toBe(true);
  });
});

describe("pileRank", () => {
  it("reads the zones the way the Team board does", () => {
    expect(pileRank(card({ zone: "red" }))).toBeLessThan(pileRank(card({ zone: "yellow" })));
    expect(pileRank(card({ zone: "yellow" }))).toBeLessThan(pileRank(card({ zone: "gray" })));
    expect(pileRank(card({ zone: "gray" }))).toBeLessThan(pileRank(card({ zone: "green" })));
  });

  it("puts a debt before everything, whatever kind of work it is", () => {
    // It was due and is not done; that outranks what sort of thing it is.
    expect(pileRank(card({ overdue: true, zone: "green" }))).toBeLessThan(
      pileRank(card({ zone: "red" })),
    );
  });

  it("puts the project's own work after the debts and before the zones", () => {
    const slot = card({ epic: "Storage" });
    expect(pileRank(slot)).toBeGreaterThan(pileRank(card({ overdue: true })));
    expect(pileRank(slot)).toBeLessThan(pileRank(card({ zone: "red" })));
  });

  // A process turn is a commitment made elsewhere too, and it will take the
  // week's time whether or not anybody plans around it — so it reads right
  // after the project's work and before anything a zone decides. Already
  // filed or only drawn ahead makes no difference to the week it costs.
  it("puts the processes' work right after the project's", () => {
    const turn = card({ task: "t1", zone: "green" });
    const ghost = { ...card({ zone: "green" }), projected: true };
    for (const c of [turn, ghost]) {
      expect(pileRank(c)).toBeGreaterThan(pileRank(card({ epic: "Storage" })));
      expect(pileRank(c)).toBeLessThan(pileRank(card({ zone: "red" })));
    }
  });

  it("leaves a card of no zone at the back", () => {
    expect(pileRank(card())).toBeGreaterThan(pileRank(card({ zone: "green" })));
  });

  it("sorts a cell without disturbing what stands equal", () => {
    // Two cards of one zone keep the order the board holds them in, so an
    // order somebody set by hand still means something among its peers.
    const pile = [
      card({ itemId: "green", zone: "green" }),
      card({ itemId: "planned-1", zone: "gray" }),
      card({ itemId: "late", overdue: true, zone: "green" }),
      card({ itemId: "planned-2", zone: "gray" }),
      card({ itemId: "urgent", zone: "red" }),
    ];
    expect([...pile].sort(byPile((c) => c)).map((c) => c.itemId)).toEqual([
      "late",
      "urgent",
      "planned-1",
      "planned-2",
      "green",
    ]);
  });
});

describe("orderWith", () => {
  const cell = ["a", "b", "c"];

  it("puts a newcomer where the pointer says", () => {
    expect(orderWith(cell, "x", 0)).toEqual(["x", "a", "b", "c"]);
    expect(orderWith(cell, "x", 2)).toEqual(["a", "b", "x", "c"]);
  });

  it("puts one dropped past the end at the end", () => {
    expect(orderWith(cell, "x", 9)).toEqual(["a", "b", "c", "x"]);
  });

  it("takes a card out before putting it back, so dragging DOWN lands where the pointer is", () => {
    // The off-by-one this exists for: "a" dragged to the third place must
    // end up third, not second — the reader is aiming at a list that no
    // longer holds the card they are carrying.
    expect(orderWith(cell, "a", 2)).toEqual(["b", "c", "a"]);
  });

  it("and dragging UP lands there too", () => {
    expect(orderWith(cell, "c", 0)).toEqual(["c", "a", "b"]);
  });

  it("leaves the order alone when the card is put back where it was", () => {
    expect(orderWith(cell, "b", 1)).toEqual(cell);
  });

  it("makes a list of one out of an empty cell", () => {
    expect(orderWith([], "x", 3)).toEqual(["x"]);
  });
});

describe("anchorFor", () => {
  const cell = ["a", "b", "c"];

  it("names the card before it", () => {
    expect(anchorFor(cell, "c")).toEqual({ after: "b" });
  });

  it("names the card now second when there is nothing before it", () => {
    // Nothing to sit after at the top, so the write says what it sits
    // before — the card that took the second place.
    expect(anchorFor(cell, "a")).toEqual({ before: "b" });
  });

  it("has nothing to say about a card alone in its cell", () => {
    expect(anchorFor(["a"], "a")).toBeNull();
  });

  it("has nothing to say about a card that is not there", () => {
    expect(anchorFor(cell, "x")).toBeNull();
  });
});

describe("placedIn", () => {
  it("is nothing for a parked card, whatever week it carries", () => {
    // The two are exclusive and every door that gives a week takes the card
    // off its shelf — but the storage is a git repository anything may write
    // to, so a card CAN arrive carrying both. Drawn by its week it would
    // stand in the grid AND in the drawer: the same work twice, counted
    // twice against the week. Mirrors board.TriageWeekOf.
    expect(placedIn(card({ week: "2026-09-07", parked: true }))).toBeNull();
  });

  it("is the card's week, and nothing else", () => {
    expect(placedIn(card({ week: "2026-08-31" }))).toBe("2026-08-31");
  });

  it("is nothing at all for a card nobody has dated", () => {
    expect(placedIn(card())).toBeNull();
  });
});

describe("weeksCovered", () => {
  it("is the one week for a card that was never stretched", () => {
    expect(weeksCovered(card({ week: "2026-08-31" }))).toEqual(["2026-08-31"]);
  });

  it("runs through the week the end date reaches", () => {
    expect(weeksCovered(card({ week: "2026-08-31", day: "2026-09-11" }))).toEqual([
      "2026-08-31",
      "2026-09-07",
    ]);
  });

  it("reaches nowhere on an end date inside its own week", () => {
    expect(weeksCovered(card({ week: "2026-08-31", day: "2026-09-04" }))).toEqual(["2026-08-31"]);
  });

  it("covers nothing at all without a week", () => {
    expect(weeksCovered(card({ day: "2026-09-11" }))).toEqual([]);
    expect(reachOf(card())).toBe("");
  });
});

// A card placed in a week AHEAD is on no day board until its Monday: that is
// what makes the backlog a regulator rather than a list.
describe("placed ahead", () => {
  const TODAY = "2026-09-03";

  it("is a week after this one", () => {
    expect(placedAhead(card({ week: "2026-09-07" }), TODAY)).toBe(true);
  });

  it("is not this week, nor a week already past", () => {
    expect(placedAhead(card({ week: "2026-08-31" }), TODAY)).toBe(false);
    expect(placedAhead(card({ week: "2026-08-24" }), TODAY)).toBe(false);
  });

  it("is not a card with no week at all", () => {
    expect(placedAhead(card({}), TODAY)).toBe(false);
  });
});

// Deferring is the act of taking a card off the board until a later day, so
// its week must not keep drawing it today: a card pushed a month out went on
// standing in this week's grid, on a board the person had just cleared it
// from. Mirrors board.deferred.
describe("deferred", () => {
  it("says a card scheduled for a later day is off today's board", () => {
    expect(deferred({ startDate: "2026-10-03" }, "2026-09-03")).toBe(true);
  });

  it("says nothing of a card standing on today, or on a day gone by", () => {
    expect(deferred({ startDate: "2026-09-03" }, "2026-09-03")).toBe(false);
    expect(deferred({ startDate: "2026-08-20" }, "2026-09-03")).toBe(false);
  });

  it("says nothing of a card with no date at all — a week is not a deferral", () => {
    expect(deferred({}, "2026-09-03")).toBe(false);
  });
});

// A card whose days ran out is drawn only on the Triage grid, in the week it
// was owed in. Dragging it into the week the team is working is the gesture
// that says "this is being done now", so it has to come back with days, or the
// drag moves the column and nothing else — which is where nine process turns
// sat on the production board, three of them a month past their week. Mirrors
// boardservice.Place / daysRanOut.
describe("brought back into the week being worked", () => {
  const TODAY = "2026-09-09";
  const WEEK = "2026-09-07";

  it("re-dates a card whose days are all behind this week", () => {
    expect(
      broughtBack({ startDate: "2026-08-24", day: "2026-08-30" }, WEEK, TODAY),
    ).toEqual({ startDate: TODAY, day: "2026-09-13" });
  });

  it("starts it TODAY, not on the week's Monday — a start before the sprint began joins the previous one", () => {
    const got = broughtBack({ startDate: "2026-08-24", day: "2026-08-30" }, WEEK, TODAY);
    expect(got?.startDate).toBe(TODAY);
    expect(got?.startDate).not.toBe(WEEK);
  });

  it("leaves a range that still reaches this week alone", () => {
    expect(
      broughtBack({ startDate: "2026-08-31", day: "2026-09-09" }, WEEK, TODAY),
    ).toBeNull();
    // the boundary: a card whose last day IS the week's Monday still stands
    expect(
      broughtBack({ startDate: "2026-08-31", day: WEEK }, WEEK, TODAY),
    ).toBeNull();
  });

  it("leaves a card with no days at all — that one joins the sprint instead", () => {
    expect(broughtBack({}, WEEK, TODAY)).toBeNull();
  });

  it("never re-dates a project slot: its dates are its row", () => {
    expect(
      broughtBack({ startDate: "2026-08-24", day: "2026-08-30", epic: "Auth" }, WEEK, TODAY),
    ).toBeNull();
  });

  it("says nothing about a week that is not the one being worked", () => {
    expect(
      broughtBack({ startDate: "2026-08-24", day: "2026-08-30" }, "2026-08-31", TODAY),
    ).toBeNull();
  });
});

// Mirrors board.TestAReviewStandsInTheWeekItsDatesFallIn. A review has no
// week of its own — the week belongs to the card it reviews — but it is work
// in the reviewer's hands and counts in their load, and drawn by nothing it
// stood on no board at all once its dates ran out.
describe("where a review card stands", () => {
  it("stands in the week its own dates fall in", () => {
    expect(placedIn({ reviewOf: "orig", startDate: "2026-07-23", day: "2026-07-23" } as Card)).toBe(
      "2026-07-20",
    );
    expect(placedIn({ reviewOf: "orig", day: "2026-09-09" } as Card)).toBe("2026-09-07");
  });

  it("keeps a week of its own, and invents none from nothing", () => {
    expect(placedIn({ reviewOf: "orig", week: "2026-09-07", startDate: "2026-07-23" } as Card)).toBe(
      "2026-09-07",
    );
    expect(placedIn({ reviewOf: "orig" } as Card)).toBeNull();
  });

  it("leaves an ordinary dated card in no column: dates are not a week", () => {
    expect(placedIn({ startDate: "2026-07-23", day: "2026-07-23" } as Card)).toBeNull();
  });
});

// A review is not one of the boxes stacked in a cell — it is folded into the
// line at the cell's foot — so it takes no part in that cell's manual order.
// Counting it made the list of ids and the boxes on screen disagree, and a
// drop then wrote "before" a card the reader could not see.
describe("what takes part in a cell's order", () => {
  it("leaves review cards out, and everything else in", () => {
    expect(ordersWithinCell({ reviewOf: "orig" } as Card)).toBe(false);
    expect(ordersWithinCell({} as Card)).toBe(true);
    expect(ordersWithinCell({ reviewOf: undefined } as Card)).toBe(true);
  });
});

// The reading order of a week's pile is about DEBTS — a card whose day has
// passed and is still open — while the late MARK is about a promise made on
// another board (board.Owed vs board.Overdue). They part company on the
// ordinary card: it is a debt in the current week's column and it is not
// painted late, so the pile must not read the mark to find it.
describe("a debt reads first, whether or not it is marked late", () => {
  const THIS = "2026-09-07";
  it("puts a card owed in an earlier week at the top", () => {
    const debt = card({ week: "2026-08-31", zone: "green" });
    expect(pileRank(debt, THIS)).toBe(0);
    // Without a week to compare against, it is just its zone again.
    expect(pileRank(debt)).toBeGreaterThan(0);
  });

  it("leaves this week's own work in its zone order", () => {
    expect(pileRank(card({ week: THIS, zone: "red" }), THIS)).toBeGreaterThan(0);
    expect(pileRank(card({ week: "2026-09-14", zone: "red" }), THIS)).toBeGreaterThan(0);
  });

  it("still puts a card the server marked late first", () => {
    expect(pileRank(card({ overdue: true, week: THIS, zone: "green" }), THIS)).toBe(0);
  });
});

File: pkg/board/triage_test.go
package board

import "testing"

// NeedsTriage is the question the strip asks: has anybody said WHEN this
// card's work is due? Everything below is a reason the answer is "nobody has
// to" — the card is not the reader's decision to make.
func TestNeedsTriageIsACardNobodyPlaced(t *testing.T) {
	t.Parallel()
	open := Card{ItemID: "c1", Title: "A card", Team: "core"}

	for _, tc := range []struct {
		name string
		card Card
		want bool
	}{
		{"an open card of its own, with no week", open, true},
		{"one merely under way is still unplanned", with(open, func(c *Card) { c.Progress = 40 }), true},
		{
			// The day's planning put it there, not the week's: how long the
			// work takes is still nobody's answer.
			"a day on the board is not an answer",
			with(open, func(c *Card) { c.Day = "2026-09-02" }),
			true,
		},
		{"a card that already has a week", with(open, func(c *Card) { c.Week = "2026-08-31" }), false},
		{"a subtask, which follows its parent", with(open, func(c *Card) { c.Parent = "c9" }), false},
		{
			"a review card, which follows the card it reviews",
			with(open, func(c *Card) { c.ReviewOf = "c9" }),
			false,
		},
		{
			// It is waiting on a reviewer, not on a week: asking for one
			// would be asking the reader to decide something already decided.
			"a card sent to review",
			with(open, func(c *Card) { c.Stage = StageReview; c.Progress = 85 }),
			false,
		},
		{"a card on a personal board", with(open, func(c *Card) { c.Domain = "~kvaps" }), false},
		{"work already finished", with(open, func(c *Card) { c.Stage = StageDone }), false},
		{"work at a hundred percent", with(open, func(c *Card) { c.Progress = 100 }), false},
		{"a state card the board keeps for itself", with(open, func(c *Card) { c.Title = SprintStateTitle }), false},
		{
			// Locked is not done and not in review: somebody still has to say
			// when it is due.
			"a locked card",
			with(open, func(c *Card) { c.Stage = StageLocked }),
			true,
		},
	} {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			if got := NeedsTriage(Board{}, tc.card, "2026-09-02"); got != tc.want {
				t.Fatalf("NeedsTriage = %v, want %v", got, tc.want)
			}
		})
	}
}

// B1. A card placed in a week AHEAD is on no day board until its Monday —
// that is what makes the strip a regulator rather than a list. The rule has
// three readers (the Team grid, the Me view, and what a carry-over counts as
// being carried now), so it is pinned here on its own rather than only
// through them.
func TestPlacedAheadIsOffEveryDayBoardUntilItsMonday(t *testing.T) {
	today := "2026-09-03" // a Thursday; its week begins 2026-08-31
	for _, tc := range []struct {
		name string
		week string
		want bool
	}{
		{"the week after this one", "2026-09-07", true},
		{"a month out", "2026-10-05", true},
		// The boundary the rule turns on: this Monday is NOT ahead, so the
		// card is on the board the moment its week begins.
		{"this very week", "2026-08-31", false},
		{"a week gone by, still owed", "2026-08-24", false},
		{"no week at all — it is in the strip, not ahead", "", false},
	} {
		t.Run(tc.name, func(t *testing.T) {
			if got := PlacedAhead(Card{Week: tc.week}, today); got != tc.want {
				t.Fatalf("PlacedAhead(week %q) = %v, want %v", tc.week, got, tc.want)
			}
		})
	}
}

func with(c Card, f func(*Card)) Card {
	f(&c)
	return c
}

// A REVIEW card has no week of its own — the week belongs to the card it
// reviews — but it is real work in the reviewer's hands, and until now it
// stood on no board at all once its dates ran out: not the day boards (the
// dates are past), not the strip (nobody is waiting on a week for it), not
// the grid (no week). It still counted in the reviewer's load, which is how
// a person with two cards in front of them read as carrying four.
//
// So when a board asks to see reviews, a review stands in the week its own
// DATES fall in — and one whose week has gone comes into the current column
// by the same debt rule as everything else.
func TestAReviewStandsInTheWeekItsDatesFallIn(t *testing.T) {
	b := Board{}
	today := "2026-09-08"
	r := Card{ItemID: "r", ReviewOf: "orig", StartDate: "2026-07-23", Day: "2026-07-23"}
	if got := TriageWeekOf(b, r, today); got != "2026-07-20" {
		t.Errorf("a review's column = %q, want the Monday of its own dates", got)
	}
	// Its end date is what it is drawn by when it has no start — a review
	// mirrors the day of the card it reviews and may carry either.
	if got := TriageWeekOf(b, Card{ReviewOf: "o", Day: "2026-09-09"}, today); got != "2026-09-07" {
		t.Errorf("a review dated only by its end = %q, want 2026-09-07", got)
	}
	// A week of its own — only a direct write to the repository makes one —
	// still wins: it is the more explicit statement of the two.
	withWeek := Card{ReviewOf: "o", Week: "2026-09-07", StartDate: "2026-07-23"}
	if got := TriageWeekOf(b, withWeek, today); got != "2026-09-07" {
		t.Errorf("a review carrying a week = %q, want that week", got)
	}
	// Nothing to place it by is no column: it is not invented from nothing.
	if got := TriageWeekOf(b, Card{ReviewOf: "o"}, today); got != "" {
		t.Errorf("an undated review = %q, want no column", got)
	}
	// And it is never in the STRIP. The strip asks its reader to say WHEN,
	// and nobody is waiting on that for a review.
	if NeedsTriage(b, r, today) {
		t.Error("a review must not stand in the triage strip")
	}
	// An ordinary card is untouched: dates are not a week, and a card nobody
	// gave a week to belongs in the strip, not in the column its day lands in.
	if got := TriageWeekOf(b, Card{StartDate: "2026-07-23", Day: "2026-07-23"}, today); got != "" {
		t.Errorf("an ordinary dated card = %q, want no column", got)
	}
}

// WORK THAT WAS DONE IN A WEEK IS THAT WEEK'S WORK, whether or not anybody
// planned it. A card closed today without ever being given a week stood in no
// column and in no strip — the strip is for work still to be looked at, and
// this work is finished — so it was on the Triage board nowhere at all, and a
// lead reading the week saw less than the week had done.
//
// It stands in the week it was FINISHED in. That is where the reader is
// looking for it, it is the week whose points it is part of (a finished card
// WITH a week already counts there), and a card finished long ago falls
// outside the window like anything else of that week rather than piling into
// the current one.
func TestFinishedWorkNobodyPlannedStandsInTheWeekItWasDoneIn(t *testing.T) {
	var b Board
	done := Card{ItemID: "done", Team: "t", Progress: 100, DoneAt: "2026-09-09",
		StartDate: "2026-09-09", Day: "2026-09-09"}
	if got := TriageWeekOf(b, done, "2026-09-09"); got != "2026-09-07" {
		t.Errorf("the week it was finished in = %q, want its Monday", got)
	}
	// Still open, still unplaced: the strip, which is where it is asked
	// about — no week is invented for work nobody has looked at.
	open := done
	open.Progress, open.DoneAt = 40, ""
	if got := TriageWeekOf(b, open, "2026-09-09"); got != "" {
		t.Errorf("open work nobody placed stands in no column, got %q", got)
	}
	if !NeedsTriage(b, open, "2026-09-09") {
		t.Error("and it is in the strip")
	}
	// A week somebody DID give it wins: the plan is what they said, and the
	// day it was finished does not move the card out of it.
	placed := done
	placed.Week = "2026-08-31"
	if got := TriageWeekOf(b, placed, "2026-09-09"); got != "2026-08-31" {
		t.Errorf("the week somebody gave it = %q", got)
	}
	// Finished work is never in the strip, placed or not.
	if NeedsTriage(b, done, "2026-09-09") {
		t.Error("finished work is not waiting to be looked at")
	}
	// And a card that recorded no day is where it always was: nowhere. There
	// is nothing to place it by.
	silent := done
	silent.DoneAt = ""
	if got := TriageWeekOf(b, silent, "2026-09-09"); got != "" {
		t.Errorf("nothing says which week finished it, got %q", got)
	}
}
